src/script/tari_script.rs

```diff
@@ -47,26 +52,76 @@ macro_rules! script {
 #[derive(Clone, Debug, PartialEq, Eq)]
 pub struct TariScript {
     script: Vec<Opcode>,
+    state: ScriptState,
 }
 
 impl TariScript {
     pub fn new(script: Vec<Opcode>) -> Self {
-        TariScript { script }
+        TariScript {
+            script,
+            state: ScriptState::default(),
+        }
     }
 
-    /// Executes the script using a default context
-    pub fn execute(&self, inputs: &ExecutionStack) -> Result<(), ScriptError> {
+    /// Executes the script using a default context. If successful, returns the final stack item.
+    pub fn execute(&mut self, inputs: &ExecutionStack) -> Result<StackItem, ScriptError> {
         self.execute_with_context(inputs, &ScriptContext::default())
     }
 
-    /// Execute the script with the given inputs and the provided context
-    pub fn execute_with_context(&self, inputs: &ExecutionStack, context: &ScriptContext) -> Result<(), ScriptError> {
+    /// Execute the script with the given inputs and the provided context. If successful, returns the final stack item.
+    pub fn execute_with_context(
+        &mut self,
+        inputs: &ExecutionStack,
+        context: &ScriptContext,
+    ) -> Result<StackItem, ScriptError>
+    {
         // Copy all inputs onto the stack
         let mut stack = inputs.clone();
-        for opcode in self.script.iter() {
-            self.execute_opcode(opcode, &mut stack, context)?
+
+        let script = self.script.clone();
```

Comment:
Perhaps `state` should rather be created as a local var here and then lives only as long as the execution